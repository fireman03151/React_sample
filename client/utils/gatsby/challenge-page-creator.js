const path = require('path');
const { dasherize } = require('../../../utils/slugs');
const { sortChallengeFiles } = require('../../../utils/sort-challengefiles');
const { challengeTypes, viewTypes } = require('../challenge-types');

const backend = path.resolve(
  __dirname,
  '../../src/templates/Challenges/projects/backend/show.tsx'
);
const classic = path.resolve(
  __dirname,
  '../../src/templates/Challenges/classic/show.tsx'
);
const frontend = path.resolve(
  __dirname,
  '../../src/templates/Challenges/projects/frontend/show.tsx'
);
const codeAlly = path.resolve(
  __dirname,
  '../../src/templates/Challenges/codeally/show.tsx'
);
const intro = path.resolve(
  __dirname,
  '../../src/templates/Introduction/intro.tsx'
);
const superBlockIntro = path.resolve(
  __dirname,
  '../../src/templates/Introduction/super-block-intro.tsx'
);
const video = path.resolve(
  __dirname,
  '../../src/templates/Challenges/video/show.tsx'
);

const odin = path.resolve(
  __dirname,
  '../../src/templates/Challenges/odin/show.tsx'
);

const views = {
  backend,
  classic,
  modern: classic,
  frontend,
  video,
  codeAlly,
  odin
  // quiz: Quiz
};

function getIsFirstStep(_node, index, nodeArray) {
  const current = nodeArray[index];
  const previous = nodeArray[index - 1];

  if (!previous) return true;
  return previous.node.challenge.block !== current.node.challenge.block;
}

function getNextChallengePath(_node, index, nodeArray) {
  const next = nodeArray[index + 1];
  return next ? next.node.challenge.fields.slug : null;
}

function getPrevChallengePath(_node, index, nodeArray) {
  const prev = nodeArray[index - 1];
  return prev ? prev.node.challenge.fields.slug : null;
}

function getTemplateComponent(challengeType) {
  return views[viewTypes[challengeType]];
}

exports.createChallengePages = function (createPage) {
  return function ({ node: { challenge } }, index, allChallengeEdges) {
    const {
      certification,
      superBlock,
      block,
      fields: { slug },
      required = [],
      template,
      challengeType,
      id
    } = challenge;
    // TODO: challengeType === 7 and isPrivate are the same, right? If so, we
    // should remove one of them.

    createPage({
      path: slug,
      component: getTemplateComponent(challengeType),
      context: {
        challengeMeta: {
          certification,
          superBlock,
          block,
          isFirstStep: getIsFirstStep(challenge, index, allChallengeEdges),
          template,
          required,
          nextChallengePath: getNextChallengePath(
            challenge,
            index,
            allChallengeEdges
          ),
          prevChallengePath: getPrevChallengePath(
            challenge,
            index,
            allChallengeEdges
          ),
          id
        },
        projectPreview: getProjectPreviewConfig(challenge, allChallengeEdges),
        slug
      }
    });
  };
};

function getProjectPreviewConfig(challenge, allChallengeEdges) {
  const { block, challengeOrder, challengeType, usesMultifileEditor } =
    challenge;

  const challengesInBlock = allChallengeEdges
    .filter(({ node: { challenge } }) => challenge.block === block)
    .map(({ node: { challenge } }) => challenge);
  const lastChallenge = challengesInBlock[challengesInBlock.length - 1];
  const solutionToLastChallenge = sortChallengeFiles(
    lastChallenge.solutions[0] ?? []
  );
  const lastChallengeFiles = sortChallengeFiles(
    lastChallenge.challengeFiles ?? []
  );
  const projectPreviewChallengeFiles = lastChallengeFiles.map((file, id) => ({
    ...file,
    contents: solutionToLastChallenge[id]?.contents ?? file.contents
  }));

  return {
    showProjectPreview:
      challengeOrder === 0 &&
      usesMultifileEditor &&
      challengeType !== challengeTypes.multifileCertProject,
    challengeData: {
      challengeType: lastChallenge.challengeType,
      challengeFiles: projectPreviewChallengeFiles
    }
  };
}

exports.createBlockIntroPages = function (createPage) {
  return function (edge) {
    const {
      fields: { slug },
      frontmatter: { block }
    } = edge.node;

    createPage({
      path: slug,
      component: intro,
      context: {
        block: dasherize(block),
        slug
      }
    });
  };
};

exports.createSuperBlockIntroPages = function (createPage) {
  return function (edge) {
    const {
      fields: { slug },
      frontmatter: { superBlock, certification }
    } = edge.node;

    if (!certification) {
      throw Error(
        `superBlockIntro page, '${superBlock}' must have certification in frontmatter`
      );
    }

    // TODO: throw if it encounters an unknown certification. Also, handle
    // coding-interview-prep. it's not a certification, but it is a superBlock.

    createPage({
      path: slug,
      component: superBlockIntro,
      context: {
        certification,
        superBlock,
        slug
      }
    });
  };
};
