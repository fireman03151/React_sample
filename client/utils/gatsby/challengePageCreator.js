const path = require('path');
const { dasherize } = require('../../../utils/slugs');

const { viewTypes } = require('../challengeTypes');

const backend = path.resolve(
  __dirname,
  '../../src/templates/Challenges/projects/backend/Show.js'
);
const classic = path.resolve(
  __dirname,
  '../../src/templates/Challenges/classic/Show.js'
);
const frontend = path.resolve(
  __dirname,
  '../../src/templates/Challenges/projects/frontend/Show.js'
);
const intro = path.resolve(
  __dirname,
  '../../src/templates/Introduction/Intro.js'
);
const superBlockIntro = path.resolve(
  __dirname,
  '../../src/templates/Introduction/SuperBlockIntro.js'
);
const video = path.resolve(
  __dirname,
  '../../src/templates/Challenges/video/Show.js'
);

const views = {
  backend,
  classic,
  modern: classic,
  frontend,
  video
  // quiz: Quiz
};

const getNextChallengePath = (node, index, nodeArray) => {
  const next = nodeArray[index + 1];
  return next ? next.node.fields.slug : '/learn';
};

const getPrevChallengePath = (node, index, nodeArray) => {
  const prev = nodeArray[index - 1];
  return prev ? prev.node.fields.slug : '/learn';
};

const getTemplateComponent = challengeType => views[viewTypes[challengeType]];

const getIntroIfRequired = (node, index, nodeArray) => {
  const next = nodeArray[index + 1];
  const isEndOfBlock = next && next.node.challengeOrder === 0;
  let nextSuperBlock = '';
  let nextBlock = '';
  if (next) {
    const { superBlock, block } = next.node;
    nextSuperBlock = superBlock;
    nextBlock = block;
  }
  return isEndOfBlock
    ? `/learn/${dasherize(nextSuperBlock)}/${dasherize(nextBlock)}`
    : '';
};

exports.createChallengePages = createPage => ({ node }, index, thisArray) => {
  const {
    superBlock,
    block,
    fields: { slug },
    required = [],
    template,
    challengeType,
    id
  } = node;
  // TODO: challengeType === 7 and isPrivate are the same, right? If so, we
  // should remove one of them.

  return createPage({
    path: slug,
    component: getTemplateComponent(challengeType),
    context: {
      challengeMeta: {
        superBlock,
        block: block,
        introPath: getIntroIfRequired(node, index, thisArray),
        template,
        required,
        nextChallengePath: getNextChallengePath(node, index, thisArray),
        prevChallengePath: getPrevChallengePath(node, index, thisArray),
        id
      },
      slug
    }
  });
};

exports.createBlockIntroPages = createPage => edge => {
  const {
    fields: { slug },
    frontmatter: { block }
  } = edge.node;

  return createPage({
    path: slug,
    component: intro,
    context: {
      block: dasherize(block),
      slug
    }
  });
};

exports.createSuperBlockIntroPages = createPage => edge => {
  const {
    fields: { slug },
    frontmatter: { superBlock }
  } = edge.node;

  return createPage({
    path: slug,
    component: superBlockIntro,
    context: {
      superBlock: superBlock,
      slug
    }
  });
};
