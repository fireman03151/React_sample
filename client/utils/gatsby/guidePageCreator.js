const path = require('path');
const select = require('unist-util-select');
const { head } = require('lodash');

const { isAStubRE } = require('../regEx');

const guideArticle = path.resolve(
  __dirname,
  '../../src/templates/Guide/GuideArticle.js'
);

exports.createGuideArticlePages = createPage => ({
  node: {
    htmlAst,
    excerpt,
    fields: { slug, component },
    id
  }
}) => {
  let meta = {};

  if (!isAStubRE.test(excerpt)) {
    const featureImage = head(select(htmlAst, 'element[tagName=img]'));
    meta.featureImage = featureImage
      ? featureImage.properties.src
      : 'https://s3.amazonaws.com/freecodecamp' +
        '/freecodecamp-square-logo-large.jpg';

    const description = head(select(htmlAst, 'element[tagName=p]'));
    meta.description = description ? description.children[0].value : '';
  }

  return createPage({
    path: `/guide${slug}`,
    component: !component
      ? guideArticle
      : path.resolve(
          __dirname,
          '../../src/templates/Guide/components/',
          component
        ),
    context: {
      id,
      meta
    }
  });
};
