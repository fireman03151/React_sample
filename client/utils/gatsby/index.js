const challengePageCreators = require('./challengePageCreator');
const guidePageCreators = require('./guidePageCreator');
const newsPageCreators = require('./newsPageCreator');

module.exports = {
  ...challengePageCreators,
  ...guidePageCreators,
  ...newsPageCreators
};
