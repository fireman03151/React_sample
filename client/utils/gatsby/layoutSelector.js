import React from 'react';
import PropTypes from 'prop-types';

import {
  CertificationLayout,
  DefaultLayout
} from '../../src/components/layouts';
import FourOhFourPage from '../../src/pages/404';

export default function layoutSelector({ element, props }) {
  const {
    location: { pathname }
  } = props;

  if (element.type === FourOhFourPage) {
    return <DefaultLayout pathname={pathname}>{element}</DefaultLayout>;
  }
  if (/^\/certification(\/.*)*/.test(pathname)) {
    return (
      <CertificationLayout pathname={pathname}>{element}</CertificationLayout>
    );
  }
  if (/^\/guide(\/.*)*/.test(pathname)) {
    console.log('Hitting guide for some reason. Need a redirect.');
  }

  const splitPath = pathname.split('/');
  const splitPathThree = splitPath.length > 2 ? splitPath[3] : '';

  const isNotSuperBlockIntro =
    splitPath.length > 3 && splitPathThree.length > 1;

  if (/^\/learn(\/.*)*/.test(pathname) && isNotSuperBlockIntro) {
    return (
      <DefaultLayout pathname={pathname} showFooter={false}>
        {element}
      </DefaultLayout>
    );
  }
  if (/^\/donation(\/.*)*|^\/$|^\/donate(\/.*)*/.test(pathname)) {
    return (
      <DefaultLayout pathname={pathname} useTheme={false}>
        {element}
      </DefaultLayout>
    );
  }
  return <DefaultLayout pathname={pathname}>{element}</DefaultLayout>;
}

layoutSelector.propTypes = {
  element: PropTypes.any,
  location: PropTypes.objectOf({ pathname: PropTypes.string }),
  props: PropTypes.any
};
