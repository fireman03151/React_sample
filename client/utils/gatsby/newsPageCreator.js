const path = require('path');

const newsArticle = path.resolve(
  __dirname, '../../src/templates/News/ShowArticle/index.js'
);

exports.createNewsArticle = createPage => ({
  node: {
    fields: { slug },
    shortId,
    id
  }
}) =>
  createPage({
    path: slug,
    component: newsArticle,
    context: {
      slug,
      shortId,
      id
    }
  });
