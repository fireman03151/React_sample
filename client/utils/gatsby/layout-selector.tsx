import React from 'react';

import CertificationLayout from '../../src/components/layouts/certification';
import DefaultLayout from '../../src/components/layouts/default';
import FourOhFourPage from '../../src/pages/404';
import { isChallenge } from '../../src/utils/path-parsers';

interface LayoutSelectorProps {
  element: JSX.Element;
  props: {
    location: { pathname: string };
    pageContext?: { challengeMeta?: { block?: string; superBlock?: string } };
  };
}
export default function layoutSelector({
  element,
  props
}: LayoutSelectorProps): JSX.Element {
  const {
    location: { pathname }
  } = props;

  if (element.type === FourOhFourPage) {
    return (
      <DefaultLayout pathname={pathname} showFooter={true}>
        {element}
      </DefaultLayout>
    );
  } else if (/\/certification\//.test(pathname)) {
    return (
      <CertificationLayout pathname={pathname}>{element}</CertificationLayout>
    );
  } else if (isChallenge(pathname)) {
    return (
      <DefaultLayout
        pathname={pathname}
        showFooter={false}
        isChallenge={true}
        block={props.pageContext?.challengeMeta?.block}
        superBlock={props.pageContext?.challengeMeta?.superBlock}
      >
        {element}
      </DefaultLayout>
    );
  } else {
    return (
      <DefaultLayout pathname={pathname} showFooter={true}>
        {element}
      </DefaultLayout>
    );
  }
}
