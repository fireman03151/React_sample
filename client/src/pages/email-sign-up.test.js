import { navigate } from 'gatsby';
import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import EmailSignUp from './email-sign-up';

const middlewares = [];
const mockStore = configureStore(middlewares);

jest.mock('../analytics');

jest.mock('gatsby', () => ({
  navigate: jest.fn()
}));

describe('<EmailSignUp />', () => {
  let component;
  let tree;

  describe('Non-Authenticated user "not accepted terms and condition"', () => {
    beforeEach(() => {
      const initialState = {
        app: {
          appUsername: '',
          user: {},
          userFetchState: {
            pending: false,
            complete: true,
            errored: false,
            error: null
          }
        }
      };

      const store = mockStore(initialState);
      const container = renderer.create(
        <Provider store={store}>
          <EmailSignUp />
        </Provider>
      );

      component = container.root;
      tree = container.toJSON();
    });

    it('it should render correctly', () => {
      expect(tree).toMatchSnapshot();
    });

    it('it should have a intro description section', () => {
      // eslint-disable-next-line testing-library/await-async-query
      const sectionIntroDescription = component.findByProps({
        className: 'intro-description'
      });

      expect(sectionIntroDescription).toBeTruthy();
    });
  });

  describe('Authenticated user "accepted terms and condition"', () => {
    beforeEach(() => {
      const initialState = {
        app: {
          appUsername: 'John Doe',
          user: {
            'John Doe': {
              acceptedPrivacyTerms: true,
              name: 'John Doe'
            }
          },
          userFetchState: {
            pending: false,
            complete: true,
            errored: false,
            error: null
          }
        }
      };
      const store = mockStore(initialState);
      store.dispatch = jest.fn();

      const container = renderer.create(
        <Provider store={store}>
          <EmailSignUp />
        </Provider>
      );

      component = container.root;
      tree = container.toJSON();
    });

    it('it should render correctly', () => {
      expect(tree).toMatchSnapshot();
    });

    it('it should redirect to learn url', () => {
      expect(navigate).toHaveBeenCalledWith('/learn');
    });
  });
});
