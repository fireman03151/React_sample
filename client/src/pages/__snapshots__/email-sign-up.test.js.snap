// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EmailSignUp /> Authenticated user "accepted terms and condition" it should render correctly 1`] = `null`;

exports[`<EmailSignUp /> Non-Authenticated user "not accepted terms and condition" it should render correctly 1`] = `
<div
  className="container"
>
  <div
    className="row"
  >
    <div
      className="col-md-8 col-md-offset-2 col-sm-10 col-sm-offset-1 col-xs-12"
    >
      <div
        className="spacer"
        style={
          {
            "height": "1px",
            "padding": "15px 0",
          }
        }
      />
      <div
        className="intro-description"
      >
        <strong>
          learn.read-this.heading
        </strong>
        <div
          className="spacer"
          style={
            {
              "height": "1px",
              "padding": "15px 0",
            }
          }
        />
        <p>
          learn.read-this.p1
        </p>
        <p>
          learn.read-this.p2
        </p>
        <p>
          learn.read-this.p3
        </p>
        <p>
          learn.read-this.p4
        </p>
        <p>
          learn.read-this.p5
        </p>
        <p>
          learn.read-this.p6
        </p>
        <p>
          learn.read-this.p7
        </p>
        <p>
          learn.read-this.p8
        </p>
        <p>
          <a
            className="inline"
            href="https://youtube.com/freecodecamp"
            rel="noopener noreferrer"
            target="_blank"
          />
        </p>
        <p>
          learn.read-this.p10
        </p>
        <p>
          <a
            className="inline"
            href="https://forum.freecodecamp.org"
            rel="noopener noreferrer"
            target="_blank"
          />
        </p>
        <p>
          learn.read-this.p12
        </p>
      </div>
      <hr />
    </div>
  </div>
  <div
    className="email-sign-up row"
    data-cy="email-sign-up"
  >
    <div
      className="col-md-8 col-md-offset-2 col-sm-10 col-sm-offset-1 col-xs-12"
    >
      <strong>
        misc.quincy
      </strong>
      <div
        className="spacer"
        style={
          {
            "height": "1px",
            "padding": "15px 0",
          }
        }
      />
      <p>
        misc.email-blast
      </p>
      <div
        className="spacer"
        style={
          {
            "height": "1px",
            "padding": "15px 0",
          }
        }
      />
    </div>
    <div
      className="col-md-8 col-md-offset-2 col-sm-10 col-sm-offset-1 col-xs-12"
    >
      <div
        className="spacer"
        style={
          {
            "height": "1px",
            "padding": "5px 0",
          }
        }
      />
      <a
        className="big-cta-btn btn btn-lg btn-primary btn-block"
        href="http://localhost:3000/signin"
        onClick={[Function]}
        onKeyDown={[Function]}
      >
        buttons.sign-up-email-list
      </a>
      <div
        className="spacer"
        style={
          {
            "height": "1px",
            "padding": "5px 0",
          }
        }
      />
    </div>
    <div
      className="col-xs-12"
    >
      <div
        className="spacer"
        style={
          {
            "height": "1px",
            "padding": "15px 0",
          }
        }
      />
    </div>
  </div>
</div>
`;
